File: core/src/test/scala/unit/kafka/log/remote/RemoteLogIndexEntryTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.log.remote

import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
import java.nio.charset.StandardCharsets

import kafka.utils.TestUtils
import org.junit.Assert.assertEquals
import org.junit.Test

class RemoteLogIndexEntryTest {
  @Test
  def testWriteAndReadInputStream(): Unit = {
    val entries = List(
      RemoteLogIndexEntry(0, 99, 0, 999, 50, "aaa".getBytes(StandardCharsets.UTF_8)),
      RemoteLogIndexEntry(100, 199, 1000, 1999, 50, "bbb".getBytes(StandardCharsets.UTF_8)),
      RemoteLogIndexEntry(200, 299, 2000, 2999, 50, "ccc".getBytes(StandardCharsets.UTF_8))
    )

    val buffers = entries.map(_.asBuffer)

    // It's OK to not close the streams, they're ephemeral.
    val outputStream = new ByteArrayOutputStream(buffers.map(_.limit()).sum)
    buffers.foreach(b => outputStream.write(b.array()))

    val inputStream = new ByteArrayInputStream(outputStream.toByteArray)
    val readEntries = RemoteLogIndexEntry.readAll(inputStream)
    assertEquals(entries, readEntries)
  }

  @Test
  def testWriteAndReadFileChannel(): Unit = {
    val entries = List(
      RemoteLogIndexEntry(0, 99, 0, 999, 50, "aaa".getBytes(StandardCharsets.UTF_8)),
      RemoteLogIndexEntry(100, 199, 1000, 1999, 50, "bbb".getBytes(StandardCharsets.UTF_8)),
      RemoteLogIndexEntry(200, 299, 2000, 2999, 50, "ccc".getBytes(StandardCharsets.UTF_8))
    )

    val channel = TestUtils.tempChannel()
    entries.foreach(e => channel.write(e.asBuffer))
    channel.position(0)

    val readEntries = RemoteLogIndexEntry.readAll(channel)
    assertEquals(entries, readEntries)
  }
}

File: clients/src/test/java/org/apache/kafka/common/log/remote/storage/RemoteLogIndexEntryTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.common.log.remote.storage;

import org.apache.kafka.test.TestUtils;
import org.junit.Assert;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.charset.StandardCharsets;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

class RemoteLogIndexEntryTest {
    @Test
    public void testWriteAndReadInputStream() throws Exception {
        List<RemoteLogIndexEntry> entries = Arrays.asList(
                new RemoteLogIndexEntry((short) 0, 99, 0, 999, 50, 100, 20, "aaa".getBytes(StandardCharsets.UTF_8)),
                new RemoteLogIndexEntry((short) 100, 199, 1000, 1999, 50, 100, 20, "bbb".getBytes(StandardCharsets.UTF_8)),
                new RemoteLogIndexEntry((short) 200, 299, 2000, 2999, 50, 100, 20, "ccc".getBytes(StandardCharsets.UTF_8))
        );

        Stream<ByteBuffer> buffers = entries.stream().map(RemoteLogIndexEntry::asBuffer);

        // It's OK to not close the streams, they're ephemeral.
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream(buffers.map(byteBuffer -> byteBuffer.limit()).mapToInt(x -> x).sum());
        buffers.forEach(b -> {
            try {
                outputStream.write(b.array());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });

        ByteArrayInputStream inputStream = new ByteArrayInputStream(outputStream.toByteArray());
        List<RemoteLogIndexEntry> readEntries = RemoteLogIndexEntry.readAll(inputStream);
        Assert.assertEquals(entries, readEntries);
    }

    @Test
    public void testWriteAndReadFileChannel() throws Exception {
        List<RemoteLogIndexEntry> entries = Arrays.asList(
                new RemoteLogIndexEntry((short) 0, 99, 0, 999, 50, 100, 20, "aaa".getBytes(StandardCharsets.UTF_8)),
                new RemoteLogIndexEntry((short) 100, 199, 1000, 1999, 50, 100, 20, "bbb".getBytes(StandardCharsets.UTF_8)),
                new RemoteLogIndexEntry((short) 200, 299, 2000, 2999, 50, 100, 20, "ccc".getBytes(StandardCharsets.UTF_8))
        );


        FileChannel channel = FileChannel.open(TestUtils.tempFile().toPath(), StandardOpenOption.READ, StandardOpenOption.WRITE);
        entries.forEach(e -> {
            try {
                channel.write(e.asBuffer());
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });
        channel.position(0);

        List<RemoteLogIndexEntry> readEntries = RemoteLogIndexEntry.readAll(channel);
        Assert.assertEquals(entries, readEntries);
    }
}
